### PR TITLE
Use QgsExtentWidget on Raster and Mesh Calculator 

### LINKED ISSUE #45247: Harmonize raster and mesh calculators GUI: improve the extent selector options
### What is the bug or the crash?

It could be nice to harmonize the two dialogs since they have a lot in common

Align the extent selection widgets. I can't find any reason we couldn't use another layer to set the extent in the raster calculator widget. But instead of going the "mask layer" route, as in the mesh calculator dialog, it could be better to move both to the "Extent" widget (as seen in the export raster layer dialog) to fill the extent coordinates.
image

image

### Steps to reproduce the issue

np

### Versions

master 3.21

### Supported QGIS version

- [ ] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ismailsunni**
## Description

Fix #45247 : use QgsExtentWidget on Raster Calculator
Peek 2022-02-09 20-39 - extent in raster and mesh calculator

Notes:
- The resolution and CRS will be updated when the user selects the extent from the layer list (similar to the previous behavior)
- I added the **Map Canvas Extent** button because it also exists in the Save Raster dialog (mentioned in the issue)
- I add the **Current Layer Extent** button because it exists in the Save Raster dialog
- I didn't add the **Draw on Canvas** button because it's not in the Save Raster dialog.
- Suggestions are welcome for those buttons (should be removed or added)

~I will continue for the Mesh Calculator after this one gets merged (and no one has worked on it)~

Funded by Camptocamp https://www.camptocamp.com

### CODE REVIEW

**Comment by nyalldawson** [comment]
Frozen till post 3.24.0 release

**Comment by ismailsunni** [comment]
Hi, @nyalldawson do you think it can be merged now since we have already passed the 3.24 release date? Thanks!

**Comment by nyalldawson** [comment]
> I add the Current Layer Extent button because it exists in the Save Raster dialog

What does this actually do in this context? It's in the Save Raster dialog because there's a singular source layer in that context, but here there's many potential referenced layers and I can't see what "current layer" would actually refer to.

**Comment by ismailsunni** [comment]
> What does this actually do in this context? It's in the Save Raster dialog because there's a singular source layer in that context, but here there's many potential referenced layers and I can't see what "current layer" would actually refer to

@nyalldawson, it refers to the currently selected layer. If the currently selected layer is not a raster layer, the button will not be shown. For example below the current layer is a vector, thus the button is not shown.

image

**Comment by nyalldawson** [comment]
@ismailsunni 
> it refers to the currently selected layer.

IMO I'd just remove this option. It's too confusing, and if a user needs to match a specific layer's extent they've already got that ability through the Layer drop down menu option in the widget.

**Comment by ismailsunni** [comment]
@nyalldawson Done.

File: src/gui/qgsextentwidget.h [hunk 1]
```diff
@@ -236,6 +236,11 @@ class GUI_EXPORT QgsExtentWidget : public QWidget, private Ui::QgsExtentGroupBox
      */
     void toggleDialogVisibility( bool visible );
 
+    /**
+     * Emitted when the layer selected is changed
+     */
```
Comment (nyalldawson) [inline src/gui/qgsextentwidget.h:241]:
```suggestion
     * \since QGIS 3.26
     */
```

File: src/gui/qgsextentgroupbox.h [hunk 2]
```diff
@@ -211,6 +212,12 @@ class GUI_EXPORT QgsExtentGroupBox : public QgsCollapsibleGroupBox
      */
     void extentChanged( const QgsRectangle &r );
 
+    /**
+     * Emitted when the layer selected is changed
+     */
```
Comment (nyalldawson) [inline src/gui/qgsextentgroupbox.h:217]:
```suggestion
     * \since QGIS 3.26
     */
```

File: src/gui/qgsextentgroupbox.h [hunk 3]
```diff
@@ -21,6 +21,7 @@
 
 #include "qgscoordinatereferencesystem.h"
 #include "qgsrectangle.h"
+#include "qgsmaplayer.h"
```
Comment (nyalldawson) [inline src/gui/qgsextentgroupbox.h:24]:
```suggestion
```

This should be forward declared instead:

class QgsMapLayer

File: src/app/mesh/qgsmeshcalculatordialog.h [hunk 4]
```diff
@@ -34,15 +34,19 @@ class APP_EXPORT QgsMeshCalculatorDialog: public QDialog, private Ui::QgsMeshCal
      * \param parent widget
      * \param f window flags
      */
-    QgsMeshCalculatorDialog( QgsMeshLayer *meshLayer = nullptr, QWidget *parent = nullptr, Qt::WindowFlags f = Qt::WindowFlags() );
+    QgsMeshCalculatorDialog(
+      const QgsRectangle &currentExtent,
+      const QgsCoordinateReferenceSystem &currentCrs,
+      QgsMeshLayer *meshLayer = nullptr,
+      QWidget *parent = nullptr,
+      Qt::WindowFlags f = Qt::WindowFlags() );
```
Comment (nyalldawson) [inline src/app/mesh/qgsmeshcalculatordialog.h:42]: How about:

```suggestion
    QgsMeshCalculatorDialog(
      QgsMeshLayer *meshLayer = nullptr,
      QgsMapCanvas* mapCanvas = nullptr,
      QWidget *parent = nullptr,
      Qt::WindowFlags f = Qt::WindowFlags() );
```

instead, and then retrieving the extent/crs from the canvas instead the dialog constructor. That's a cleaner approach in my view (also for the raster calculator)

Comment (ismailsunni) [reply src/app/mesh/qgsmeshcalculatordialog.h]: @nyalldawson done! Thanks for the review

**Comment by ismailsunni** [comment]
Hi @nyalldawson, is this PR ready to be merged or is there something that I need to address first? Thanks!

**Comment by DelazJ** [comment]
This PR seems to fix a report and from what I understand @nyalldawson 's requests are addressed. Can we revive it? @ismailsunni If ever you do not have rights to reopen, I can, after you rebased. Thanks

**Comment by ismailsunni** [comment]
@DelazJ I am happy to rebase and you can reopen it later (I don't have the right). I need to fix the build first though, it's been a while since I build QGIS.
Thanks @DelazJ

**Comment by ismailsunni** [comment]
hi @DelazJ 
I just update my branch for this PR. But it seems, the PR is not updated. Maybe because it's closed? Should I create a new PR or could you re-open this and maybe the PR's branch is updated?

Sorry for sooo late update.
Thanks!

**Comment by ismailsunni** [comment]
hi @DelazJ, I am sorry, it seems no one is re-reviewing again the PR. Perhaps it was a holiday season or the PR is not good enough to be merged.

**Pull request closed without merging**